Sources/Plasma/Apps/plClient/Mac-Cocoa/main.mm

```diff
@@ -0,0 +1,542 @@
+/*==LICENSE==*
+
+CyanWorlds.com Engine - MMOG client, server and tools
+Copyright (C) 2011  Cyan Worlds, Inc.
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+Additional permissions under GNU GPL version 3 section 7
+
+If you modify this Program, or any covered work, by linking or
+combining it with any of RAD Game Tools Bink SDK, Autodesk 3ds Max SDK,
+NVIDIA PhysX SDK, Microsoft DirectX SDK, OpenSSL library, Independent
+JPEG Group JPEG library, Microsoft Windows Media SDK, or Apple QuickTime SDK
+(or a modified version of those libraries),
+containing parts covered by the terms of the Bink SDK EULA, 3ds Max EULA,
+PhysX SDK EULA, DirectX SDK EULA, OpenSSL and SSLeay licenses, IJG
+JPEG Library README, Windows Media SDK EULA, or QuickTime SDK EULA, the
+licensors of this Program grant you additional
+permission to convey the resulting work. Corresponding Source for a
+non-source form of such a combination shall include the source code for
+the parts of OpenSSL and IJG JPEG Library used as well as that of the covered
+work.
+
+You can contact Cyan Worlds, Inc. by email legal@cyan.com
+ or by snail mail at:
+      Cyan Worlds, Inc.
+      14617 N Newport Hwy
+      Mead, WA   99021
+
+*==LICENSE==*/
+
+#import "PLSKeyboardEventMonitor.h"
+#include "plClient/plClient.h"
+#include "plClient/plClientLoader.h"
+#include "plCmdParser.h"
+#include "plProduct.h"
+#if PLASMA_PIPELINE_GL
+#include "pfGLPipeline/plGLPipeline.h"
+#endif
+#include <algorithm>
+#include <regex>
+#include <unordered_set>
+
+#include "PLSPatcherWindowController.h"
+#include "pfConsoleCore/pfConsoleEngine.h"
+#include "pfGameGUIMgr/pfGameGUIMgr.h"
+#include "plInputCore/plInputDevice.h"
+#include "plNetClient/plNetClientMgr.h"
+#include "plNetGameLib/plNetGameLib.h"
+
+#include "PLSLoginWindowController.h"
+#include "PLSServerStatus.h"
+
+#import "Cocoa/Cocoa.h"
+#if PLASMA_PIPELINE_GL
+#import <OpenGL/gl.h>
+#endif
```

Comment:
I don't think this is necessary? libepoxy and/or pfGLPipeline is supposed to handle this sort of thing for us

Reply:
I need to go look at exactly why this include is here. My hunch is that PLSView might need this. But it could probably be refined further.

The client is responsible for setting up the render layer, which means it needs some awareness of the current engine. The window also has the awareness for what display it is on - and what GPU it needs to render to. So the client will have to match the window ID back to the GPU. That might only require CGL - I'd have to look.

Reply:
I'm going to remove this import. There are a few reasons this is here - but arguably that code may be in the wrong spot.

One reason is that this code queries the renderer name to populate the window title on debug builds. That code isn't here right now - but it does exist in the Metal branch. The Metal branch no longer makes the right GL calls - but it should at some point. Not essential, could move.

On the Metal side - [main.mm is responsible for pairing the correct Metal device with the current window and fetching the next frame buffer](https://github.com/colincornaby/Plasma/blob/b0b95679f004779bac2f8fb7fd7309ae15192823/Sources/Plasma/Apps/plClient/Mac-Cocoa/main.mm#L438).

In OpenGL - you're going to have similar issues. In the "modern" and performant OpenGL system, you can't just bait into a view. You've got to wait for the layer to fire a callback for you, and the layer will hand you a context. So just as Metal has to do a handshake for the current frame buffer - you'll have to do a handshake for the current context. (I'm using "modern" in quotes here because layer based rendering was introduced in 10.5, so it should be everywhere we're thinking about.) My impression from conversations has been that directly blitting to views from OpenGL is highly discouraged - especially on modern macOS, and may have bugs.

This code may be in the wrong place - and might be suffering a bit from the "everything in main.mm" strategy that was inherited from Windows. I think this could should actually go into PLSView. That doesn't need to be dealt with here in since that code doesn't exist in this branch.

There are going to be several things to watch for when implementing GL:
- On some dual GPU systems - like Intel MacBook Pros - the current GPU will change mid-flight. This is because in response to our launch the system will changeover from the Intel GPU to the AMD/Nvidia one. This bit us on some MacBook Pros because the frame buffer ended up on one GPU but resources ended up on another. This is why Metal rechecks the current GPU every frame. There probably is a better way of doing this... but might be another reason it's important to adopt modern layer based rendering on macOS. macOS will manage the current context for you.
- Metal doesn't have the context problem because it approaches the GPU as a generic compute device. So I can render any time, any where. But I do need to get a frame buffer from the window server to render into. The lambda I linked to is my solution from drilling back up from the engine into the layer. I think Plasma has a built in structure for passing views and windows down into the renderer. It might be easier to approach passing the layer or render buffer through as a view.
- 

-

Reply:
Cleaned up the imports. There is a lot of stuff being imported into main because of the unfortunate everything in main thing going on. That's pretty atypical for Cocoa apps. Hopefully as stuff gets factored into individual classes these imports in main can be streamlined.